File: frontend/src/services/adapters/purchasing.js
const optionalText = (value) => value?.trim() || null

export function mapSupplierFromApi(supplier) {
  return {
    id: supplier.id,
    name: supplier.name,
    rnc: supplier.rnc || '',
    contactName: supplier.contactName || '',
    phone: supplier.phone || '',
    email: supplier.email || '',
    address: supplier.address || '',
    branchIds: supplier.branchIds || [],
    productCount: Number(supplier.productCount) || 0,
    active: supplier.active,
    version: supplier.version,
    createdAt: supplier.createdAt,
    updatedAt: supplier.updatedAt,
    apiSynced: true,
  }
}

export function supplierToApiPayload(form) {
  if (!form.name?.trim()) throw new Error('Ingresa el nombre del proveedor.')
  if (!form.branchIds?.length) throw new Error('Selecciona al menos una sucursal autorizada.')
  return {
    name: form.name.trim(),
    rnc: optionalText(form.rnc),
    contactName: optionalText(form.contactName),
    phone: optionalText(form.phone),
    email: optionalText(form.email),
    address: optionalText(form.address),
    branchIds: form.branchIds,
  }
}

export function mapPurchaseRequestFromApi(request) {
  return {
    id: request.id,
    number: request.number,
    supplierId: request.supplierId,
    supplierName: request.supplierName,
    branchId: request.branchId,
    requesterName: request.requesterName,
    requesterId: request.requesterId,
    items: (request.items || []).map((item) => ({
      id: item.id,
      name: item.name,
      qty: Number(item.qty) || 0,
      unit: item.unit,
      price: Number(item.price) || 0,
      subtotal: Number(item.subtotal) || 0,
    })),
    status: request.status,
    priority: request.priority,
    notes: request.notes || '',
    quoteFile: request.quoteFile,
    total: Number(request.total) || 0,
    createdAt: request.createdAt,
    reviewedAt: request.reviewedAt,
    reviewedBy: request.reviewedBy,
    deliveredAt: request.deliveredAt,
    version: request.version,
    updatedAt: request.updatedAt,
    apiSynced: true,
  }
}

export function purchaseRequestToApiPayload(form) {
  if (!form.supplierId) throw new Error('Selecciona un proveedor.')
  if (!form.branchId) throw new Error('Selecciona una sucursal.')
  const items = (form.items || []).filter((item) => item.name?.trim())
  if (!items.length) throw new Error('Agrega al menos un artículo.')
  return {
    supplierId: form.supplierId,
    branchId: form.branchId,
    items: items.map((item) => ({
      name: item.name.trim(),
      qty: Number(item.qty),
      unit: item.unit?.trim() || 'unidad',
      price: Number(item.price),
    })),
    priority: form.priority || 'normal',
    notes: optionalText(form.notes),
    quoteFile: form.quoteFile?.name ? { name: form.quoteFile.name } : null,
  }
}

export function mapPurchasingSettingsFromApi(settings) {
  return {
    approverUserId: settings.approverUserId || '',
    approverUser: settings.approverUser || null,
    notifyOnRequest: settings.notifyOnRequest,
    version: settings.version,
    updatedAt: settings.updatedAt,
    apiSynced: true,
  }
}

export function purchasingSettingsToApiPayload(settings, version) {
  return {
    version,
    approverUserId: settings.approverUserId || null,
    notifyOnRequest: Boolean(settings.notifyOnRequest),
  }
}

File: backend/app/schemas/purchasing.py
from datetime import datetime
from decimal import Decimal
from typing import Annotated, Literal
from uuid import UUID

from pydantic import EmailStr, Field, StringConstraints, field_validator, model_validator

from app.schemas.common import ApiModel

SupplierSortField = Literal["name", "rnc", "contactName", "productCount", "createdAt"]
PurchaseRequestSortField = Literal[
    "createdAt", "number", "supplier", "requester", "total", "status", "priority"
]
SortDirection = Literal["asc", "desc"]
PurchaseRequestStatus = Literal["pendiente", "aprobada", "rechazada", "entregada"]
PurchaseRequestPriority = Literal["normal", "alta"]

RequiredText = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]


def _optional_text(value: str | None) -> str | None:
    if value is None:
        return None
    normalized = value.strip()
    return normalized or None


class SupplierResponse(ApiModel):
    id: UUID
    name: str
    rnc: str | None
    contact_name: str | None
    phone: str | None
    email: str | None
    address: str | None
    branch_ids: list[UUID]
    product_count: int
    active: bool
    version: int
    created_at: datetime
    updated_at: datetime


class PaginatedSuppliersResponse(ApiModel):
    items: list[SupplierResponse]
    page: int
    page_size: int
    total_items: int
    total_pages: int


class CreateSupplierRequest(ApiModel):
    name: RequiredText = Field(max_length=200)
    rnc: str | None = Field(default=None, max_length=80)
    contact_name: str | None = Field(default=None, max_length=160)
    phone: str | None = Field(default=None, max_length=40)
    email: EmailStr | None = None
    address: str | None = Field(default=None, max_length=500)
    branch_ids: list[UUID] = Field(min_length=1, max_length=100)

    _normalize_optional_fields = field_validator(
        "rnc", "contact_name", "phone", "address", mode="before"
    )(_optional_text)

    @field_validator("branch_ids")
    @classmethod
    def unique_branches(cls, value: list[UUID]) -> list[UUID]:
        if len(value) != len(set(value)):
            raise ValueError("No repitas sucursales autorizadas.")
        return value


class UpdateSupplierRequest(ApiModel):
    version: int = Field(ge=1)
    name: RequiredText | None = Field(default=None, max_length=200)
    rnc: str | None = Field(default=None, max_length=80)
    contact_name: str | None = Field(default=None, max_length=160)
    phone: str | None = Field(default=None, max_length=40)
    email: EmailStr | None = None
    address: str | None = Field(default=None, max_length=500)
    branch_ids: list[UUID] | None = Field(default=None, min_length=1, max_length=100)
    active: bool | None = None

    _normalize_optional_fields = field_validator(
        "rnc", "contact_name", "phone", "address", mode="before"
    )(_optional_text)

    @field_validator("branch_ids")
    @classmethod
    def unique_branches(cls, value: list[UUID] | None) -> list[UUID] | None:
        if value is not None and len(value) != len(set(value)):
            raise ValueError("No repitas sucursales autorizadas.")
        return value

    @model_validator(mode="after")
    def require_change(self) -> UpdateSupplierRequest:
        if not (set(self.model_fields_set) - {"version"}):
            raise ValueError("Indica al menos un campo para actualizar.")
        return self


class PurchaseQuoteFile(ApiModel):
    name: RequiredText = Field(max_length=255)


class PurchaseRequestItemInput(ApiModel):
    name: RequiredText = Field(max_length=240)
    qty: Decimal = Field(gt=0, max_digits=14, decimal_places=3)
    unit: RequiredText = Field(max_length=40)
    price: Decimal = Field(ge=0, max_digits=14, decimal_places=2)


class PurchaseRequestItemResponse(ApiModel):
    id: UUID
    name: str
    qty: Decimal
    unit: str
    price: Decimal
    subtotal: Decimal


class PurchaseRequestResponse(ApiModel):
    id: UUID
    number: str
    supplier_id: UUID
    supplier_name: str
    branch_id: UUID
    requester_name: str
    requester_id: UUID
    items: list[PurchaseRequestItemResponse]
    status: PurchaseRequestStatus
    priority: PurchaseRequestPriority
    notes: str | None
    quote_file: PurchaseQuoteFile | None
    total: Decimal
    created_at: datetime
    reviewed_at: datetime | None
    reviewed_by: UUID | None
    delivered_at: datetime | None
    version: int
    updated_at: datetime


class PaginatedPurchaseRequestsResponse(ApiModel):
    items: list[PurchaseRequestResponse]
    page: int
    page_size: int
    total_items: int
    total_pages: int


class PurchaseRequestStatsResponse(ApiModel):
    total: int
    pendiente: int
    aprobada: int
    rechazada: int
    entregada: int


class CreatePurchaseRequestRequest(ApiModel):
    supplier_id: UUID
    branch_id: UUID
    items: list[PurchaseRequestItemInput] = Field(min_length=1, max_length=100)
    priority: PurchaseRequestPriority = "normal"
    notes: str | None = Field(default=None, max_length=2000)
    quote_file: PurchaseQuoteFile | None = None

    _normalize_notes = field_validator("notes", mode="before")(_optional_text)


class UpdatePurchaseRequestRequest(ApiModel):
    version: int = Field(ge=1)
    supplier_id: UUID | None = None
    branch_id: UUID | None = None
    items: list[PurchaseRequestItemInput] | None = Field(
        default=None, min_length=1, max_length=100
    )
    priority: PurchaseRequestPriority | None = None
    notes: str | None = Field(default=None, max_length=2000)
    quote_file: PurchaseQuoteFile | None = None

    _normalize_notes = field_validator("notes", mode="before")(_optional_text)

    @model_validator(mode="after")
    def require_change(self) -> UpdatePurchaseRequestRequest:
        if not (set(self.model_fields_set) - {"version"}):
            raise ValueError("Indica al menos un campo para actualizar.")
        return self


class ReviewPurchaseRequestRequest(ApiModel):
    version: int = Field(ge=1)
    status: Literal["aprobada", "rechazada"]


class DeliverPurchaseRequestRequest(ApiModel):
    version: int = Field(ge=1)


class PurchasingApproverResponse(ApiModel):
    id: UUID
    name: str


class PurchasingSettingsResponse(ApiModel):
    approver_user_id: UUID | None
    approver_user: PurchasingApproverResponse | None
    notify_on_request: bool
    version: int
    updated_at: datetime


class UpdatePurchasingSettingsRequest(ApiModel):
    version: int = Field(ge=1)
    approver_user_id: UUID | None = None
    notify_on_request: bool
